docs/src/main/paradox/java/http/routing-dsl/directives/index.md

```diff
@@ -118,3 +130,31 @@ is in fact the most readable one.
 
 Still, the purpose of the exercise presented here is to show you how flexible directives can be and how you can
 use their power to define your web service behavior at the level of abstraction that is right for **your** application.
+
+## Type Safety of Directives
+
+When you combine directives with `anyOf` and `allOf` methods the routing DSL makes sure that all extractions work as
+expected and logical constraints are enforced at compile-time.
+
+For example you cannot call `anyOf`  with a directive producing an extraction with one that doesn't:
+
+```java
+anyOf(this::get, this::extractClientIP, routeProvider) // doesn't compile
+```
+
+Also the number of extractions and their types have to match up:
```

Comment:
please explain why we're using partial, it can be rather confusing at first look :)